rustls/src/compress.rs

```diff
@@ -75,3 +81,67 @@ pub struct DecompressionFailed;
 /// A content-less error for when `CertCompressor::compress` fails.
 #[derive(Debug)]
 pub struct CompressionFailed;
+
+#[cfg(feature = "zlib-rs")]
+mod feat_zlib_rs {
+    use zlib_rs::{deflate, inflate, ReturnCode};
+
+    use super::*;
+
+    /// A certificate decompressor for the Zlib algorithm using the `zlib-rs` crate.
+    #[derive(Debug)]
+    pub struct ZlibRsDecompressor;
+
+    impl CertDecompressor for ZlibRsDecompressor {
+        fn algorithm(&self) -> CertificateCompressionAlgorithm {
+            CertificateCompressionAlgorithm::Zlib
+        }
+
+        fn decompress(&self, input: &[u8], output: &mut [u8]) -> Result<(), DecompressionFailed> {
+            let (output_filled, rc) =
+                inflate::uncompress_slice(output, input, inflate::InflateConfig::default());
+            if rc != ReturnCode::Ok {
+                return Err(DecompressionFailed);
+            }
+
+            if output_filled.len() != output.len() {
+                return Err(DecompressionFailed);
+            }
+
+            Ok(())
+        }
+    }
+
+    /// A certificate compressor for the Zlib algorithm using the `zlib-rs` crate.
+    #[derive(Debug)]
+    pub struct ZlibRsCompressor;
+
+    impl CertCompressor for ZlibRsCompressor {
+        fn algorithm(&self) -> CertificateCompressionAlgorithm {
+            CertificateCompressionAlgorithm::Zlib
+        }
+
+        fn compress(
+            &self,
+            input: Vec<u8>,
+            level: CompressionLevel,
+        ) -> Result<Vec<u8>, CompressionFailed> {
+            let mut output = alloc::vec![0u8; deflate::compress_bound(input.len())];
+            let config = match level {
+                CompressionLevel::Interactive => deflate::DeflateConfig::default(),
+                CompressionLevel::Amortized => deflate::DeflateConfig::new(9),
```

Comment:
Is there a citation to use for this magic number? The [zlib-rs docs](https://docs.rs/zlib-rs/latest/zlib_rs/deflate/struct.DeflateConfig.html#method.new) offer no commentary, it's just a bare `i32`. Some googling around leads me to believe the effort level is 0..9 with 9 being "the most effort". Maybe sufficient to leave a comment here saying as much?

Reply:
In zlib this is `#define Z_BEST_COMPRESSION       9` and actually I can get that out of zlib-rs::c_api -- done.